In [ ]:
!pip install flaml

In [ ]:
import sys
print("현재 파이썬 버전:", sys.version)

# 현재 파이썬 인터프리터 경로에 맞춰 FLAML 재설치
!{sys.executable} -m pip install --upgrade flaml

현재 파이썬 버전: 3.12.11 (main, Jun  4 2025, 08:56:18) [GCC 11.4.0]


In [ ]:
!pip install numpy==1.26.4

In [ ]:
# 2. 라이브러리 불러오기
import pandas as pd
from sklearn.model_selection import train_test_split
from flaml import AutoML
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix, ConfusionMatrixDisplay
import matplotlib.pyplot as plt

# 3. 데이터 불러오기 (Colab 업로드 방식)
from google.colab import files
uploaded = files.upload()   # 'ObesityDataSet_raw_and_data_sinthetic.csv' 선택

# 4. CSV 파일 로드
df = pd.read_csv("ObesityDataSet_raw_and_data_sinthetic.csv")

# 5. X, y 분리
X = df.drop(columns=["NObeyesdad", "Height", "Weight"])
y = df["NObeyesdad"]

# 6. 학습/테스트 데이터 분할 (75:25, stratify=y로 클래스 균형 유지)
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.25, random_state=42, stratify=y
)

Saving ObesityDataSet_raw_and_data_sinthetic.csv to ObesityDataSet_raw_and_data_sinthetic (1).csv


In [ ]:
# 7. AutoML 설정
automl = AutoML()
settings = {
    "time_budget": 60,  # 총 탐색 시간(초). 필요 시 조정 가능
    "metric": "accuracy",  # 평가 지표
    "task": "classification",  # 분류 문제
    "log_file_name": "flaml_obesity.log",
    "seed": 42
}

# 8. AutoML 훈련
automl.fit(X_train=X_train, y_train=y_train, **settings)

[flaml.automl.logger: 09-17 05:33:51] {1752} INFO - task = classification
[flaml.automl.logger: 09-17 05:33:51] {1763} INFO - Evaluation method: cv
[flaml.automl.logger: 09-17 05:33:51] {1862} INFO - Minimizing error metric: 1-accuracy
[flaml.automl.logger: 09-17 05:33:51] {1979} INFO - List of ML learners in AutoML Run: ['lgbm', 'rf', 'xgboost', 'extra_tree', 'xgb_limitdepth', 'sgd', 'lrl1']
[flaml.automl.logger: 09-17 05:33:51] {2282} INFO - iteration 0, current learner lgbm
[flaml.automl.logger: 09-17 05:33:52] {2417} INFO - Estimated sufficient time budget=8425s. Estimated necessary time budget=195s.
[flaml.automl.logger: 09-17 05:33:52] {2466} INFO -  at 1.2s,	estimator lgbm's best error=0.4068,	best estimator lgbm's best error=0.4068
[flaml.automl.logger: 09-17 05:33:52] {2282} INFO - iteration 1, current learner lgbm
[flaml.automl.logger: 09-17 05:33:53] {2466} INFO -  at 1.6s,	estimator lgbm's best error=0.4068,	best estimator lgbm's best error=0.4068
[flaml.automl.logger: 09-1

INFO:flaml.tune.searcher.blendsearch:No low-cost partial config given to the search algorithm. For cost-frugal search, consider providing low-cost values for cost-related hps via 'low_cost_partial_config'. More info can be found at https://microsoft.github.io/FLAML/docs/FAQ#about-low_cost_partial_config-in-tune


[flaml.automl.logger: 09-17 05:33:55] {2466} INFO -  at 4.1s,	estimator sgd's best error=0.8339,	best estimator lgbm's best error=0.2925
[flaml.automl.logger: 09-17 05:33:55] {2282} INFO - iteration 5, current learner lgbm
[flaml.automl.logger: 09-17 05:33:56] {2466} INFO -  at 4.9s,	estimator lgbm's best error=0.2925,	best estimator lgbm's best error=0.2925
[flaml.automl.logger: 09-17 05:33:56] {2282} INFO - iteration 6, current learner lgbm
[flaml.automl.logger: 09-17 05:33:58] {2466} INFO -  at 6.5s,	estimator lgbm's best error=0.2925,	best estimator lgbm's best error=0.2925
[flaml.automl.logger: 09-17 05:33:58] {2282} INFO - iteration 7, current learner xgboost
[flaml.automl.logger: 09-17 05:33:59] {2466} INFO -  at 7.4s,	estimator xgboost's best error=0.4150,	best estimator lgbm's best error=0.2925
[flaml.automl.logger: 09-17 05:33:59] {2282} INFO - iteration 8, current learner lgbm
[flaml.automl.logger: 09-17 05:33:59] {2466} INFO -  at 8.3s,	estimator lgbm's best error=0.2925,	b

INFO:flaml.tune.searcher.blendsearch:No low-cost partial config given to the search algorithm. For cost-frugal search, consider providing low-cost values for cost-related hps via 'low_cost_partial_config'. More info can be found at https://microsoft.github.io/FLAML/docs/FAQ#about-low_cost_partial_config-in-tune


[flaml.automl.logger: 09-17 05:34:52] {2466} INFO -  at 60.8s,	estimator lrl1's best error=0.4643,	best estimator lgbm's best error=0.1643
[flaml.automl.logger: 09-17 05:34:52] {2724} INFO - retrain lgbm for 0.4s
[flaml.automl.logger: 09-17 05:34:52] {2727} INFO - retrained model: LGBMClassifier(learning_rate=0.21174115582543823, max_bin=15,
               min_child_samples=5, n_estimators=63, n_jobs=-1, num_leaves=53,
               reg_alpha=0.0009765625, reg_lambda=1.9018665152510525,
               verbose=-1)
[flaml.automl.logger: 09-17 05:34:52] {2009} INFO - fit succeeded
[flaml.automl.logger: 09-17 05:34:52] {2010} INFO - Time taken to find the best model: 50.89701747894287


In [ ]:
print(automl.best_config_per_estimator)

{'lgbm': {'n_estimators': 63, 'num_leaves': 53, 'min_child_samples': 5, 'learning_rate': 0.21174115582543823, 'log_max_bin': 4, 'colsample_bytree': 1.0, 'reg_alpha': 0.0009765625, 'reg_lambda': 1.9018665152510525}, 'rf': {'n_estimators': 23, 'max_features': 0.1486655975394226, 'max_leaves': 4, 'criterion': 'entropy'}, 'xgboost': {'n_estimators': 24, 'max_leaves': 9, 'min_child_weight': 1.5972787761198735, 'learning_rate': 0.25802890276133356, 'subsample': 0.7052886642806031, 'colsample_bylevel': 0.9201504037276426, 'colsample_bytree': 0.896436559310386, 'reg_alpha': 0.0009765625, 'reg_lambda': 0.8114042287963648}, 'extra_tree': {'n_estimators': 17, 'max_features': 0.21394265246709848, 'max_leaves': 6, 'criterion': 'entropy'}, 'xgb_limitdepth': {'n_estimators': 10, 'max_depth': 6, 'min_child_weight': 0.9999999999999993, 'learning_rate': 0.29999999999999993, 'subsample': 1.0, 'colsample_bylevel': 1.0, 'colsample_bytree': 1.0, 'reg_alpha': 0.0009765625, 'reg_lambda': 1.0}, 'sgd': {'penalt

In [ ]:
from sklearn.metrics import accuracy_score

results = {}

for estimator, config in automl.best_config_per_estimator.items():
    try:
        # 같은 데이터에서 개별 알고리즘 재훈련
        model = AutoML()
        model.fit(
            X_train=X_train, y_train=y_train,
            estimator_list=[estimator],  # 특정 알고리즘만 사용
            time_budget=60,              # 1분만
            metric="accuracy",
            seed=42
        )
        y_pred = model.predict(X_test)
        acc = accuracy_score(y_test, y_pred)
        results[estimator] = acc
    except Exception as e:
        print(estimator, "error:", e)

[flaml.automl.logger: 09-17 05:34:52] {1752} INFO - task = classification
[flaml.automl.logger: 09-17 05:34:52] {1763} INFO - Evaluation method: cv
[flaml.automl.logger: 09-17 05:34:52] {1862} INFO - Minimizing error metric: 1-accuracy
[flaml.automl.logger: 09-17 05:34:52] {1979} INFO - List of ML learners in AutoML Run: ['lgbm']
[flaml.automl.logger: 09-17 05:34:52] {2282} INFO - iteration 0, current learner lgbm
[flaml.automl.logger: 09-17 05:34:53] {2417} INFO - Estimated sufficient time budget=1775s. Estimated necessary time budget=2s.
[flaml.automl.logger: 09-17 05:34:53] {2466} INFO -  at 0.2s,	estimator lgbm's best error=0.4068,	best estimator lgbm's best error=0.4068
[flaml.automl.logger: 09-17 05:34:53] {2282} INFO - iteration 1, current learner lgbm
[flaml.automl.logger: 09-17 05:34:53] {2466} INFO -  at 0.4s,	estimator lgbm's best error=0.4068,	best estimator lgbm's best error=0.4068
[flaml.automl.logger: 09-17 05:34:53] {2282} INFO - iteration 2, current learner lgbm
[flaml

INFO:flaml.tune.searcher.blendsearch:No low-cost partial config given to the search algorithm. For cost-frugal search, consider providing low-cost values for cost-related hps via 'low_cost_partial_config'. More info can be found at https://microsoft.github.io/FLAML/docs/FAQ#about-low_cost_partial_config-in-tune


[flaml.automl.logger: 09-17 05:39:57] {2417} INFO - Estimated sufficient time budget=5892s. Estimated necessary time budget=6s.
[flaml.automl.logger: 09-17 05:39:57] {2466} INFO -  at 0.6s,	estimator sgd's best error=0.8339,	best estimator sgd's best error=0.8339
[flaml.automl.logger: 09-17 05:39:57] {2282} INFO - iteration 1, current learner sgd
[flaml.automl.logger: 09-17 05:39:58] {2466} INFO -  at 1.9s,	estimator sgd's best error=0.8339,	best estimator sgd's best error=0.8339
[flaml.automl.logger: 09-17 05:39:58] {2282} INFO - iteration 2, current learner sgd
[flaml.automl.logger: 09-17 05:39:59] {2466} INFO -  at 2.4s,	estimator sgd's best error=0.7094,	best estimator sgd's best error=0.7094
[flaml.automl.logger: 09-17 05:39:59] {2282} INFO - iteration 3, current learner sgd
[flaml.automl.logger: 09-17 05:39:59] {2466} INFO -  at 3.0s,	estimator sgd's best error=0.7094,	best estimator sgd's best error=0.7094
[flaml.automl.logger: 09-17 05:39:59] {2282} INFO - iteration 4, current 

INFO:flaml.tune.searcher.blendsearch:No low-cost partial config given to the search algorithm. For cost-frugal search, consider providing low-cost values for cost-related hps via 'low_cost_partial_config'. More info can be found at https://microsoft.github.io/FLAML/docs/FAQ#about-low_cost_partial_config-in-tune


[flaml.automl.logger: 09-17 05:40:57] {2417} INFO - Estimated sufficient time budget=8349s. Estimated necessary time budget=8s.
[flaml.automl.logger: 09-17 05:40:57] {2466} INFO -  at 0.9s,	estimator lrl1's best error=0.4662,	best estimator lrl1's best error=0.4662
[flaml.automl.logger: 09-17 05:40:57] {2282} INFO - iteration 1, current learner lrl1
[flaml.automl.logger: 09-17 05:40:58] {2466} INFO -  at 1.7s,	estimator lrl1's best error=0.4637,	best estimator lrl1's best error=0.4637
[flaml.automl.logger: 09-17 05:40:58] {2282} INFO - iteration 2, current learner lrl1
[flaml.automl.logger: 09-17 05:40:59] {2466} INFO -  at 2.6s,	estimator lrl1's best error=0.4637,	best estimator lrl1's best error=0.4637
[flaml.automl.logger: 09-17 05:40:59] {2282} INFO - iteration 3, current learner lrl1
[flaml.automl.logger: 09-17 05:41:00] {2466} INFO -  at 3.5s,	estimator lrl1's best error=0.4637,	best estimator lrl1's best error=0.4637
[flaml.automl.logger: 09-17 05:41:00] {2282} INFO - iteration 

In [ ]:
print("모델별 정확도:", results)

모델별 정확도: {'lgbm': 0.8276515151515151, 'rf': 0.821969696969697, 'xgboost': 0.7954545454545454, 'extra_tree': 0.7443181818181818, 'xgb_limitdepth': 0.803030303030303, 'sgd': 0.3996212121212121, 'lrl1': 0.4337121212121212}


*   LightGBM (lgbm): 0.828 → 최고 성능
*   Random Forest (rf): 0.822 → 근소한 차이
*   XGBoost (xgboost): 0.792
*   XGBoost Limit Depth (xgb_limitdepth): 0.803
*   Extra Trees (extra_tree): 0.708
*   SGD Classifier (sgd): 0.411
*   Logistic Regression L1 (lrl1): 0.434